src/store/Controller.cc

```diff
@@ -321,6 +329,26 @@ Store::Controller::hasReadableDiskEntry(const StoreEntry &e) const
     return swapDir->hasReadableEntry(e);
 }
 
+/// flags problematic ephemeral peek() results before we share them
+void
+Store::Controller::checkPeeked(const StoreEntry &entry) const
+{
+    // The "hittingRequiresCollapsing() has an active writer" checks below
+    // protect callers from getting stuck and/or from using a stale revalidation
+    // reply. However, these protections are not reliable because the writer may
+    // disappear at any time and/or without a trace. Collapsing adds risks...
+    if (entry.hittingRequiresCollapsing()) {
+        if (entry.hasTransients()) {
```

Comment:
I think that we should have already checked these has_writer-related conditions in Transients::get() - so there is no need to repeat those checks here. If so, then, basically, checkPeeked() covers only the 'local writer' case. Then I would move the 'else' block below into a separate method and call this new method from Store::Controller::peekAtLocal() - somewhere below checkTransients() call.

Reply:
> we should have already checked these has_writer-related conditions in Transients::get()

Yes and no: Yes, we should have already checked these conditions in Transients::get(), but that may have been done long time ago (and for a different HTTP request). A lot could change since then -- and that is what the "Too late" comment is talking about. Also:

* Transients::get() is not called for cache entries that already have local StoreEntry objects in store_table. 
* Failed Transients::get() is not relevant for MemStore::get() and Rock::SwapDir::get() cases -- those are handled by allowSharing() and transients->monitorIo(). There _are_ problems with the latter path, but I decided to handle them outside this PR because they are not specific to stale collapsed revalidation entries.

Finally, I speculate that we will add more checks to this new method eventually.

Thus, I think it is best to keep this method general and exclude/document impossible-to-check cases explicitly, like this PR does.

Reply:
>  but that may have been done long time ago 

I am not against additional checks (by means of Controller::checkPeeked()), but I wonder why should we call it only after 'long time has passed', trying to recalculate the context then (i.e., to understand what Store produced the StoreEntry) and finally realize that we can not complete this checking in some cases (as the comment notes). 

> Transients::get() is not called for cache entries that already have local StoreEntry objects in store_table.

Agreed, but in this case, we would do that collapsing check in Controller::peekAtLocal().

> those are handled by allowSharing() and transients->monitorIo().

This also looks a little strange - checkPeeked() is called after allowSharing(), which may have an effect on conditions, calculated in checkPeeked(). In other words, checkPeeked() may be probably renamed because it checks something produced by several methods (including peek()).

Reply:
> I wonder why should we call it only after 'long time has passed', trying to recalculate the context then (i.e., to understand what Store produced the StoreEntry) and finally realize that we can not complete this checking in some cases (as the comment notes).

"Long-time-after" is not exactly what is happening in this new checking code though. The new code is executed for every viable find() candidate. Some are freshly created. Some were around for a long time. If we can write a generic check that applies to both categories, we should. In this particular (SMP) case, we cannot -- we can only check freshly created entries (and we do check them, at creation time). In the non-SMP case, we can write a single/generic check. IMO, the current structure of the method and comments highlight this difference.

> in this case, we would do that collapsing check in Controller::peekAtLocal().

Yes, but peekAtLocal() may not be a good place for additional consistency checks because, besides find(), that method is used in evictIfFound() which should be rather aggressive -- if there is doubt, it should release the entry, not pretend that it does not exist like find() does.

> In other words, checkPeeked() may be probably renamed because it checks something produced by several methods (including peek()).

I agree but do not have good suggestions (unfortunately). I doubt something like checkPeekedAndPublic() or checkSharable() would be better. FWIW, I started by inlining this code, but it produced too much noise in the already non-trivial logic, so I extracted it into a method. I think this code deserves its own method, but checkPeeked() is the best name I can come up with. I know it is far from perfect. Suggestions welcomed.

Would you prefer checkFoundCandidate()?

Reply:
> not pretend that it does not exist like find() does.

Well, find() also releases doubtful entries so that distinction is misleading. We could add some checks to peekAtLocal() if needed, but if we can centralized checks, it is better to centralize them.

Reply:
> Would you prefer checkFoundCandidate()?

Or checkCandidate() - since we are already inside find() method.

> if we can centralized checks, it is better to centralize them

Ok.

Reply:
I used checkFoundCandidate() in 3c6f947. The checks (or their interpretation) kind of depends on being done in the find() context...